src/coreclr/jit/compiler.hpp

```diff
@@ -662,11 +663,19 @@ BasicBlockVisit BasicBlock::VisitAllSuccs(Compiler* comp, TFunc func)
             return VisitEHSuccs(comp, func);
 
         case BBJ_COND:
-            RETURN_ON_ABORT(func(GetFalseTarget()));
-
-            if (!TrueEdgeIs(GetFalseEdge()))
+            if (TrueEdgeIs(GetFalseEdge()))
+            {
+                RETURN_ON_ABORT(func(GetFalseTarget()));
+            }
+            else if (useProfile && (GetTrueEdge()->getLikelihood() < GetFalseEdge()->getLikelihood()))
```

Comment:
Am I confused, or is this visiting the less likely successor first?

```suggestion
            else if (useProfile && (GetTrueEdge()->getLikelihood() > GetFalseEdge()->getLikelihood()))
```

Reply:
It looks unintuitive, but I think we need to flip the comparison so the DFS is in the order we want. Consider the following block list, pre-ordering:
```
BBnum BBid ref try hnd preds           weight   IBC [IL range]   [jump]                            [EH region]        [flags]
---------------------------------------------------------------------------------------------------------------------------------------------------------------------
BB01 [0014]  1                             1        [???..???)-> BB02(1)                 (always)                     i keep internal
BB02 [0000]  1       BB01                  1    100 [000..00E)-> BB04(0),BB03(1)         ( cond )                     i IBC
BB03 [0010]  1       BB02                  0.50  50 [00D..00E)-> BB05(1)                 (always)                     i IBC nullcheck
BB04 [0011]  1       BB02                  0      0 [00D..00E)-> BB05(1)                 (always)                     i IBC rare
BB05 [0012]  2       BB03,BB04             1        [00D..017)                           (return)                     i hascall gcsafe
```

When processing `BB02`, if we visit `BB03` before `BB04`, then we end up with an RPO that looks something like `[<BB03's successors>, BB03, BB04, BB02, ...]`, so after layout, we get this:
```
BBnum BBid ref try hnd preds           weight   IBC [IL range]   [jump]                            [EH region]        [flags]
---------------------------------------------------------------------------------------------------------------------------------------------------------------------
BB01 [0014]  1                             1        [???..???)-> BB02(1)                 (always)                     i keep internal
BB02 [0000]  1       BB01                  1    100 [000..00E)-> BB04(0),BB03(1)         ( cond )                     i IBC
BB04 [0011]  1       BB02                  0      0 [00D..00E)-> BB05(1)                 (always)                     i IBC rare
BB03 [0010]  1       BB02                  0.50  50 [00D..00E)-> BB05(1)                 (always)                     i IBC nullcheck
BB05 [0012]  2       BB03,BB04             1        [00D..017)                           (return)                     i hascall gcsafe
```

If we instead visit the less likely successor (`BB04`) first, we push the more likely successor `BB03` up to `BB02` in the RPO, and get this layout:
```
BBnum BBid ref try hnd preds           weight   IBC [IL range]   [jump]                            [EH region]        [flags]
---------------------------------------------------------------------------------------------------------------------------------------------------------------------
BB01 [0014]  1                             1        [???..???)-> BB02(1)                 (always)                     i keep internal
BB02 [0000]  1       BB01                  1    100 [000..00E)-> BB04(0),BB03(1)         ( cond )                     i IBC
BB03 [0010]  1       BB02                  0.50  50 [00D..00E)-> BB05(1)                 (always)                     i IBC nullcheck
BB04 [0011]  1       BB02                  0      0 [00D..00E)-> BB05(1)                 (always)                     i IBC rare
BB05 [0012]  2       BB03,BB04             1        [00D..017)                           (return)                     i hascall gcsafe
```

Reply:
Agreed, I'll fix this in a follow-up PR. As for what the new abstraction should look like, would you prefer we move the `useProfile` check into `AllSuccessorEnumerator`, or even introduce a new enumerator like `ProfileGuidedSuccessorEnumerator`?

Reply:
Perhaps introduce two instance initializer methods on `AllSuccessorEnumerator` and then pass some factory method to `fgRunDfs`? E.g. the normal use would be:
```
fgRunDfs([](SuccessorEnumerator* enumerator, BasicBlock* block) { enumerator->InitializeAllSuccs(block); }, ...);
```
and the block layout use could be
```
fgRunDfs([](SuccessorEnumerator* enumerator, BasicBlock* block) { enumerator->InitializeAllSuccsForBlockLayout(block); }, ...);
```